src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs

```diff
@@ -406,6 +406,207 @@ public void Dispose()
             }
         }
 
+        [ThreadStatic]
+        public static int ThreadId;
+
+        [Fact]
+        public async Task GetRequiredService_UsesSingletonAndLazyLocks_NoDeadlock()
+        {
+            // Thread 1: Thing1 (transient) -> Thing0 (singleton)
+            // Thread 2: Thing2 (singleton) -> Thing1 (transient) -> Thing0 (singleton)
+
+            // 1. Thread 1 resolves the Thing1 which is a transient service
+            // 2. In parallel, Thread 2 resolves Thing2 which is a singleton
+            // 3. Thread 1 enters the factory callback for Thing1 and takes the lazy lock
+            // 4. Thread 2 takes callsite for Thing2 as a singleton lock when it resolves Thing2
+            // 5. Thread 2 enters the factory callback for Thing1 and waits on the lazy lock
+            // 6. Thread 1 calls GetRequiredService<Thing0> on the service provider, takes callsite for Thing0 causing no deadlock
+            // (rather than taking the locks that are already taken - either the lazy lock or the Thing2 callsite lock)
+
+            Thing0 thing0 = null;
+            Thing1 thing1 = null;
+            Thing2 thing2 = null;
+            IServiceProvider sp = null;
+
+            // Arrange
+            var services = new ServiceCollection();
+
+            var lazy = new Lazy<Thing1>(() =>
+            {
+                thing0 = sp.GetRequiredService<Thing0>();
+                return new Thing1(thing0);
+            });
+
+            services.AddSingleton<Thing0>();
+            services.AddTransient(sp =>
+            {
+                if (ThreadId == 1)
+                {
+                    Thread.Sleep(1000);
```

Comment:
Instead of `Thread.Sleep` calls here, can we instead pass in a Task(s) (or ManualResetEvents) that can be awaited on? Then we can control the timing outside of these threads and ensure the ordering happens as we want it.